src/sinks/nats.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{
+    config::{DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    emit,
+    event::Event,
+    internal_events::NatsEventSent,
+    sinks::util::StreamSink,
+};
+use async_trait::async_trait;
+use futures::{future, stream::BoxStream, FutureExt, StreamExt};
+use serde::{Deserialize, Serialize};
+
+/**
+ * Code dealing with the SinkConfig struct.
+ *
+ * DEV: Start with the bare minimum for now.
+ */
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct NatsSinkConfig {
+    url: String,
+    subject: String,
+}
+
+inventory::submit! {
+    SinkDescription::new::<NatsSinkConfig>("nats")
+}
+
+impl GenerateConfig for NatsSinkConfig {}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SinkConfig for NatsSinkConfig {
+    async fn build(
+        &self,
+        _cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink = NatsSink::new(self.clone());
+        let healthcheck = future::ok(()).boxed();
+
+        Ok((super::VectorSink::Stream(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+/**
+ * Code dealing with the Sink struct.
+ */
+
+#[derive(Clone)]
+pub struct NatsSink {
+    nc: nats::Connection,
+    subject: String,
+}
+
+impl NatsSink {
+    fn new(config: NatsSinkConfig) -> Self {
+        let options = nats::Options::new();
+        let nc = options.connect(&config.url).unwrap();
+
+        Self {
+            nc,
+            subject: config.subject,
+        }
+    }
+}
+
+#[async_trait]
+impl StreamSink for NatsSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            match event {
+                Event::Log(log) => {
+                    let body = log
+                        .get(crate::config::log_schema().message_key())
+                        .map(|v| v.to_string_lossy())
+                        .unwrap_or_else(|| "".into());
+
+                    let message_len = body.len();
+
+                    if self.nc.publish(&self.subject, body).is_ok() {
+                        emit!(NatsEventSent {
```

Comment:
I think we should split up these internal events into:

* `EventProcessed`: simply indicates an event was handled. This event is common accross our components. It should emit the counter, as you are, and also log at the trace level to be consistent.
* `NatsEventSendSucces`: indicates the event was successfully published. I like that you included the bytes sent.
* `NatsEventSendFail`: indicates an event failed to be sent. This should log the error (rate limited)

Reply:
Added the events, but please double check me on if I ordered their emissions correctly.

Reply:
Events look good :+1: